data/transactions/teal.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package transactions
+
+import (
+	"bytes"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// EvalDelta stores StateDeltas for an application's global key/value store, as
+// well as StateDeltas for some number of accounts holding local state for that
+// application
+type EvalDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	GlobalDelta basics.StateDelta `codec:"gd"`
+
+	// When decoding EvalDeltas, the integer key represents an offset into
+	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ...]
+	LocalDeltas map[uint64]basics.StateDelta `codec:"ld,allocbound=config.MaxEvalDeltaAccounts"`
+
+	Logs []string `codec:"lg,allocbound=config.MaxLogCalls"`
+
+	// Intentionally, temporarily wrong - need to decide how to
+	// allocbound properly when structure is recursive.  Even a bound
+	// of 2 would allow arbitrarily large object if deep.
```

Comment:
What is the solution for this one?

Reply:
It turned out that the allocbounds are not our only protection.  We don't decode streamed msgpack messages.  We take them off the wire, with bounded size, and then decode.

Reply:
Given what Nickolai said about msgpack pre-allocating an array with a decoded integer size, I think there can an amplification attack where a few decoded bytes allocate 16 new `SignedTxnWithAD` objects.

Something like this when decoding `SignedTxnWithAD`:
```
dt itx 16 dt itx 16 dt itx 16 ...
```